### PR TITLE
feat: add logical plan based prepare statement for postgresql

### PR DESCRIPTION
**Pull request by sunng87**
I hereby agree to the terms of the [GreptimeDB CLA](https://gist.github.com/xtang/6378857777706e568c1949c7578592cc)

## What's changed and what's your intention?

Use `LogicalPlan` API for parameter replacement in postgres

## Checklist

- [x]  I have written the necessary rustdoc comments.
- [x]  I have added the necessary unit tests and integration tests.

### CODE REVIEW

File: src/servers/src/postgres/types.rs [hunk 1]
```diff
@@ -0,0 +1,677 @@
...
+            .parameter::<f64>(idx)?
+            .map(|v| v.to_string())
+            .unwrap_or_else(|| "".to_owned())),
+        &Type::TIMESTAMP => Ok(portal
+            .parameter::<NaiveDateTime>(idx)?
+            .map(|v| v.format("%Y-%m-%d %H:%M:%S%.6f").to_string())
+            .unwrap_or_else(|| "".to_owned())),
+        _ => Err(PgWireError::UserError(Box::new(ErrorInfo::new(
+            "ERROR".to_owned(),
+            "22023".to_owned(),
```
Comment (MichaelScofield) [inline src/servers/src/postgres/types.rs:179]: better make "22023" a const

File: src/servers/src/postgres/types.rs [hunk 2]
```diff
@@ -0,0 +1,677 @@
...
+            Some(&format!(
+                "Expected: {}, found: {}",
+                client_param_types.len(),
+                param_count
+            )),
+        ));
+    }
+
+    for (idx, client_type) in client_param_types.iter().enumerate() {
+        if let Some(Some(server_type)) = param_types.get(&format!("${}", idx + 1)) {
```
Comment (MichaelScofield) [inline src/servers/src/postgres/types.rs:227]:
```suggestion
        let Some(Some(server_type)) = param_types.get(&format!("${}", idx + 1)) else { continue };
```
to reduce a layer of indentation

**Review by MichaelScofield — APPROVED** [review]

File: src/servers/src/postgres/types.rs [hunk 3]
```diff
@@ -0,0 +1,677 @@
...
+            .unwrap_or_else(|| "".to_owned())),
+        &Type::FLOAT8 => Ok(portal
+            .parameter::<f64>(idx)?
+            .map(|v| v.to_string())
+            .unwrap_or_else(|| "".to_owned())),
+        &Type::TIMESTAMP => Ok(portal
+            .parameter::<NaiveDateTime>(idx)?
+            .map(|v| v.format("%Y-%m-%d %H:%M:%S%.6f").to_string())
+            .unwrap_or_else(|| "".to_owned())),
+        _ => Err(PgWireError::UserError(Box::new(ErrorInfo::new(
```
Comment (killme2008) [inline src/servers/src/postgres/types.rs:177]: Can we support date/datetime/timestamp here?

File: src/servers/src/postgres/types.rs [hunk 4]
```diff
@@ -0,0 +1,677 @@
...
+                    match server_type {
+                        ConcreteDataType::Int8(_) => ScalarValue::Int8(data.map(|n| n as i8)),
+                        ConcreteDataType::Int16(_) => ScalarValue::Int16(data.map(|n| n as i16)),
+                        ConcreteDataType::Int32(_) => ScalarValue::Int32(data),
+                        ConcreteDataType::Int64(_) => ScalarValue::Int64(data.map(|n| n as i64)),
+                        ConcreteDataType::UInt8(_) => ScalarValue::UInt8(data.map(|n| n as u8)),
+                        ConcreteDataType::UInt16(_) => ScalarValue::UInt16(data.map(|n| n as u16)),
+                        ConcreteDataType::UInt32(_) => ScalarValue::UInt32(data.map(|n| n as u32)),
+                        ConcreteDataType::UInt64(_) => ScalarValue::UInt64(data.map(|n| n as u64)),
+                        ConcreteDataType::Timestamp(unit) => match *unit {
```
Comment (killme2008) [inline src/servers/src/postgres/types.rs:270]: Use below code to replace the match expr:
```rust
unit.create_timestamp(u as i64))
                    .try_to_scalar_value(t)
```

File: src/servers/src/postgres/types.rs [hunk 5]
```diff
@@ -0,0 +1,677 @@
...
+                    match server_type {
+                        ConcreteDataType::Int8(_) => ScalarValue::Int8(data.map(|n| n as i8)),
+                        ConcreteDataType::Int16(_) => ScalarValue::Int16(data.map(|n| n as i16)),
+                        ConcreteDataType::Int32(_) => ScalarValue::Int32(data.map(|n| n as i32)),
+                        ConcreteDataType::Int64(_) => ScalarValue::Int64(data),
+                        ConcreteDataType::UInt8(_) => ScalarValue::UInt8(data.map(|n| n as u8)),
+                        ConcreteDataType::UInt16(_) => ScalarValue::UInt16(data.map(|n| n as u16)),
+                        ConcreteDataType::UInt32(_) => ScalarValue::UInt32(data.map(|n| n as u32)),
+                        ConcreteDataType::UInt64(_) => ScalarValue::UInt64(data.map(|n| n as u64)),
+                        ConcreteDataType::Timestamp(unit) => match *unit {
```
Comment (killme2008) [inline src/servers/src/postgres/types.rs:309]: The same with the above suggestion.

File: src/servers/src/postgres/types.rs [hunk 6]
```diff
@@ -0,0 +1,677 @@
...
+                                    server_type, client_type
+                                )),
+                            ))
+                        }
+                    }
+                }
+                &Type::FLOAT4 => {
+                    let data = portal.parameter::<f32>(idx)?;
+                    match server_type {
+                        ConcreteDataType::Float32(_) => ScalarValue::Float32(data),
```
Comment (killme2008) [inline src/servers/src/postgres/types.rs:336]: Other integral data types such as i8,i32,u32 etc. must be able to cast into float types.

File: src/servers/src/postgres/types.rs [hunk 7]
```diff
@@ -0,0 +1,677 @@
...
+                                    "Expected: {}, found: {}",
+                                    server_type, client_type
+                                )),
+                            ))
+                        }
+                    }
+                }
+                &Type::FLOAT8 => {
+                    let data = portal.parameter::<f64>(idx)?;
+                    match server_type {
```
Comment (killme2008) [inline src/servers/src/postgres/types.rs:353]: The same as the above suggestion.

File: tests-integration/tests/sql.rs [hunk 8]
```diff
@@ -141,28 +141,37 @@ pub async fn test_postgres_crud(store_type: StorageType) {
         .await
         .unwrap();
 
-    sqlx::query("create table demo(i bigint, ts timestamp time index)")
+    sqlx::query("create table demo(i bigint, ts timestamp time index, d date)")
```
Comment (killme2008) [inline tests-integration/tests/sql.rs:144]: Test `datetime` too.

**Comment by sunng87** [comment]
resolved PTAL @killme2008

**Review by killme2008 — APPROVED** [review]
LGTM

**Pull request merged**